File: TopVotedCandidate.java
// Problem 911. Online Election
// Time Complexity : Constructor - O(N) , Query - O(logN)
// Space Complexity : O(N)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
class TopVotedCandidate {
    private HashMap<Integer, Integer> leadersMap;
    private int[] times;
    private HashMap<Integer, Integer> countMap;
    public TopVotedCandidate(int[] persons, int[] times) {
        this.times = times;
        this.leadersMap = new HashMap<>();
        this.countMap = new HashMap<>();
        int leader = -1;
        for (int i = 0; i < persons.length; i++) {
            int p = persons[i];
            int t = times[i];
            countMap.put(p, countMap.getOrDefault(p, 0) + 1);
            if (leader == -1 || countMap.get(p) >= countMap.get(leader)) {
                leader = p;
            }
            leadersMap.put(t, leader);
        }
    }
    
    public int q(int t) {
        int left = 0, right = times.length - 1;
        while (left <= right) {
            int mid = left + (right - left) / 2;
            if (times[mid] > t) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        return leadersMap.get(times[right]);
    }
}

/**
 * Your TopVotedCandidate object will be instantiated and called as such:
 * TopVotedCandidate obj = new TopVotedCandidate(persons, times);
 * int param_1 = obj.q(t);
 */

File: topVotedCandidate.cpp
// Time complexity: O (log n) where n is size of the input vector
// Space complexity: O(n) - candidate hashmap and leader at time t hashmap

/*
Use hashmaps to store the votes for each candidate and the leader at each time t

Use binary search to get the closest smallest time and return the value of the leader at that time t
*/

class TopVotedCandidate {
public:
    unordered_map<int, int> leaderAtTime;
    unordered_map<int, int> votes;
    vector<int> times;
    TopVotedCandidate(vector<int>& persons, vector<int>& times) {
        int leader = persons[0];
        this->times = times;
        for(int i = 0; i<persons.size(); i++) {
            votes[persons[i]]++;
            if(votes[persons[i]] >= votes[leader])
                leader = persons[i];
            leaderAtTime[times[i]] = leader;
        }
    }
    
    int q(int t) {
        if(leaderAtTime.find(t) != leaderAtTime.end())
            return leaderAtTime[t];
        // binary search to get the closest smaller t from hashmap 
        // the lowest closest number is found in high when the loop exits
        vector<int> inputTimes = times;
        int low = 0;
        int high = times.size()-1;
        while(low <= high) {
            int mid = (high - low)/2 + low;
            if(t > times[mid])
                low = mid + 1;
            else 
                high = mid - 1;
        }
        return leaderAtTime[times[high]];
    }
};

/**
 * Your TopVotedCandidate object will be instantiated and called as such:
 * TopVotedCandidate* obj = new TopVotedCandidate(persons, times);
 * int param_1 = obj->q(t);
 */